drivers/overlay/overlay.go

```diff
@@ -1486,3 +1595,80 @@ func nameWithSuffix(name string, number int) string {
 	}
 	return fmt.Sprintf("%s%d", name, number)
 }
+
+func (d *Driver) getAdditionalLayerPath(dgst digest.Digest, ref string) (string, error) {
+	refelem := base64.StdEncoding.EncodeToString([]byte(ref))
+	for _, ls := range d.options.layerStores {
+		var elems []string
+		if ls.withReference {
+			elems = append(elems, refelem)
+		}
+		elems = append(elems, dgst.String())
+		target := filepath.Join(append([]string{ls.path}, elems...)...)
+		// Check if all necessary files exist
+		for _, p := range []string{
+			filepath.Join(target, "diff"),
+			filepath.Join(target, "info"),
+			// TODO(ktock): We should have an API to expose the stream data of this layer
+			//              to enable the client to retrieve the entier contents of this
+			//              layer when it exports this layer.
+		} {
+			if _, err := os.Stat(p); err != nil {
+				return "", graphdriver.ErrLayerUnknown
+			}
+		}
+		return target, nil
+	}
+
+	return "", graphdriver.ErrLayerUnknown
+}
+
+type additionalLayer struct {
+	path      string
+	d         *Driver
+	closeOnce sync.Once
+}
+
+// Info returns arbitrary information stored along with this layer (i.e. `info` file).
+func (al *additionalLayer) Info() (io.ReadCloser, error) {
+	return os.Open(filepath.Join(al.path, "info"))
+}
+
+// CreateAs creates a new layer from this additional layer.
+func (al *additionalLayer) CreateAs(id, parent string) error {
+	targetDiffDir := filepath.Join(al.path, "diff")
+	if _, err := os.Stat(targetDiffDir); err != nil {
+		return err
+	}
+	if ld, err := os.Readlink(targetDiffDir); err == nil {
```

Comment:
Should we return or warn if the err!=nil?  If it does not exists?

Reply:
We have already checked that it exists above (but haven't checked permissions).

Perhaps it's better to remove `os.Stat` above and use this to return all errors except EINVAL (i.e. when the argument is not a symlink).